File: oop/04-abstraction/csharp/CreditAccount.cs
/*
 * Demonstrates: Abstraction - Concrete Implementation
 * Concept: CreditAccount implements all abstract methods from Account
 */

namespace FinanceDomain.Abstraction;

public class CreditAccount : Account
{
    public decimal CreditLimit { get; }
    public decimal InterestRate { get; }

    public CreditAccount(string accountNumber, string accountHolder, decimal creditLimit, decimal interestRate)
        : base(accountNumber, accountHolder, 0)
    {
        CreditLimit = creditLimit;
        InterestRate = interestRate;
    }

    public decimal OutstandingBalance => Math.Abs(Math.Min(_balance, 0));
    public decimal AvailableCredit => CreditLimit - OutstandingBalance;

    public override void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than zero.");

        if (amount > AvailableCredit)
            throw new InvalidOperationException($"Exceeds credit limit. Available: R{AvailableCredit:F2}");

        _balance -= amount;
    }

    public override decimal CalculateMonthlyFees()
    {
        return OutstandingBalance * InterestRate;
    }

    public override string GetAccountSummary()
    {
        return $"Credit Account {AccountNumber} | Holder: {AccountHolder} | Outstanding: R{OutstandingBalance:F2} | Available Credit: R{AvailableCredit:F2}";
    }

    public void MakePayment(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Payment amount must be greater than zero.");

        _balance += amount;
    }
}

File: oop/04-abstraction/typescript/creditAccount.ts
/*
 * Demonstrates: Abstraction - Concrete Implementation
 * Concept: CreditAccount implements all abstract methods from Account
 */

import { Account } from "./account";

export class CreditAccount extends Account {
    public readonly creditLimit: number;
    public readonly interestRate: number;

    constructor(accountNumber: string, accountHolder: string, creditLimit: number, interestRate: number) {
        super(accountNumber, accountHolder, 0);
        this.creditLimit = creditLimit;
        this.interestRate = interestRate;
    }

    get outstandingBalance(): number {
        return Math.abs(Math.min(this._balance, 0));
    }

    get availableCredit(): number {
        return this.creditLimit - this.outstandingBalance;
    }

    withdraw(amount: number): void {
        if (amount <= 0) {
            throw new Error("Amount must be greater than zero.");
        }
        if (amount > this.availableCredit) {
            throw new Error(`Exceeds credit limit. Available: R${this.availableCredit.toFixed(2)}`);
        }
        this._balance -= amount;
    }

    calculateMonthlyFees(): number {
        return this.outstandingBalance * this.interestRate;
    }

    getAccountSummary(): string {
        return `Credit Account ${this.accountNumber} | Holder: ${this.accountHolder} | Outstanding: R${this.outstandingBalance.toFixed(2)} | Available Credit: R${this.availableCredit.toFixed(2)}`;
    }

    makePayment(amount: number): void {
        if (amount <= 0) {
            throw new Error("Payment amount must be greater than zero.");
        }
        this._balance += amount;
    }
}
